### PR TITLE
✨ Add mutation to transfer study to another organization

### LINKED ISSUE #656: Allow users to be added or removed from organizations
A mutation should be added to add or remove members of an organization.

```graphql
addMember(organization: ID!, user: ID!): Organization!
```
```graphql
removeMember(organization: ID!, user: ID!): Organization!
```

### LINKED ISSUE #657: Allow studies to be transferred between organizations
We may occasionally need to move studies between organizations. For now, this should be a superuser action although in the future we may want to allow users who administrate multiple orgs to be able to transfer studies.

```graphql
transferStudy(study: ID!, organization: ID!): Organization!
```
This will move the specified study to the specified organization.

### PR DESCRIPTION
**Pull request by XuTheBunny**
✨ Add mutation to transfer study to another organization
Screen Shot 2021-06-21 at 8 38 01 PM
Screen Shot 2021-06-22 at 1 05 33 PM

Closes #657

✨ Add mutation to add member to an organization
Screen Shot 2021-06-22 at 12 12 21 PM
Screen Shot 2021-06-22 at 12 12 37 PM

✨ Add mutation to remove member from an organization
Screen Shot 2021-06-22 at 12 12 50 PM
Screen Shot 2021-06-22 at 12 13 06 PM

Closes #656

### CODE REVIEW

File: creator/studies/mutations.py [hunk 1]
```diff
@@ -736,6 +736,64 @@ def mutate(self, info, study, data):
...
+        )
+
+    study = graphene.Field(StudyNode)
+
+    def mutate(self, info, study, organization):
+        """
+        Update the organization of a study
+        """
+        user = info.context.user
+        if not user.has_perm("organizations.change_organization"):
```
Comment (dankolbman) [inline creator/studies/mutations.py:759]: We can be a little more safe and also check that the user belongs to both the source and destination organizations. 
https://github.com/kids-first/kf-api-study-creator/pull/682/files#diff-efc346caa0fbfd8a6e993133f86fd9ab49ae702faff904ad698a9ae51b73361dR74

File: creator/organizations/mutations.py [hunk 2]
```diff
@@ -113,10 +118,146 @@ def mutate(self, info, id, input):
...
+
+    organization = graphene.Field(OrganizationNode)
+
+    def mutate(self, info, organization, user):
+        """
+        Add a member to an organization
+        """
+        user_id = user
+        user = info.context.user
+        if not user.has_perm("organizations.change_organization"):
```
Comment (gsantia) [inline creator/organizations/mutations.py:143]: Maybe we want a test covering this?

File: creator/organizations/mutations.py [hunk 3]
```diff
@@ -113,10 +118,146 @@ def mutate(self, info, id, input):
...
+
+    organization = graphene.Field(OrganizationNode)
+
+    def mutate(self, info, organization, user):
+        """
+        Remove a member from an organization
+        """
+        user_id = user
+        user = info.context.user
+        if not user.has_perm("organizations.change_organization"):
```
Comment (gsantia) [inline creator/organizations/mutations.py:208]: same comment as above

File: creator/organizations/mutations.py [hunk 4]
```diff
@@ -113,10 +118,146 @@ def mutate(self, info, id, input):
...
+            raise GraphQLError("Not allowed")
+
+        try:
+            _, organization_id = from_global_id(organization)
+            organization = Organization.objects.get(pk=organization_id)
+        except Organization.DoesNotExist:
+            raise GraphQLError(
+                f"Organization {organization_id} does not exist"
```
Comment (gsantia) [inline creator/organizations/mutations.py:216]:
```suggestion
                f"Organization {organization_id} does not exist."
```
i'm sorry the other ones had a period it bugged me 😝

File: creator/organizations/mutations.py [hunk 5]
```diff
@@ -113,10 +118,146 @@ def mutate(self, info, id, input):
...
+            _, user_id = from_global_id(user_id)
+            member = User.objects.get(id=int(user_id))
+        except (User.DoesNotExist, TypeError):
+            raise GraphQLError(f"User {user_id} does not exist.")
+
+        if member.organizations.filter(pk=organization_id).exists():
+            organization.members.remove(member)
+            message = (
+                f"{user.display_name} removed {member.display_name} "
+                f"to organization {organization.name}"
```
Comment (gsantia) [inline creator/organizations/mutations.py:229]:
```suggestion
                f"from organization {organization.name}"
```

File: creator/studies/mutations.py [hunk 6]
```diff
@@ -736,6 +736,75 @@ def mutate(self, info, study, data):
...
+        if not user.organizations.filter(pk=study.organization.id).exists():
+            raise GraphQLError(
+                "Not allowed - may only transfer study from an "
+                "organization the user is a member of."
+            )
+        if not user.organizations.filter(pk=organization_id).exists():
+            raise GraphQLError(
+                "Not allowed - may only transfer study to an "
+                "organization the user is a member of."
+            )
```
Comment (gsantia) [inline creator/studies/mutations.py:785]:
```suggestion
       org_msg = (
            "Not allowed - may only transfer study from an "
             "organization the user is a member of."
        )
        if not user.organizations.filter(pk=study.organization.id).exists():
            raise GraphQLError(org_msg)
        if not user.organizations.filter(pk=organization_id).exists():
            raise GraphQLError(org_msg)
```

File: tests/studies/test_transfer_study.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+        data={"query": TRANSFER_STUDY_MUTATION, "variables": variables},
+    )
+
+    message = resp.json()["errors"][0]["message"]
+    assert "does not exist" in message
+
+
+def test_study_not_found(db, permission_client):
+    """
+    Test behavior when the specified organization does not exist
```
Comment (gsantia) [inline tests/studies/test_transfer_study.py:138]:
```suggestion
    Test behavior when the specified study does not exist
```

**Review by gsantia — COMMENTED** [review]
looks good, left some minor comments

File: creator/studies/mutations.py (hunk 6 above)
Comment (XuTheBunny) [reply creator/studies/mutations.py:785]: Here the message differences are `from an organization` and `to an organization`

**Review by dankolbman — COMMENTED** [review]
Can you update the variables to use snake case instead of camel case?

**Review by gsantia — APPROVED** [review]

**Review by dankolbman — COMMENTED** [review]
# 🥪

**Pull request merged**